### ISSUE TITLE
[8.0.0] Can no longer append visually empty rows

### ISSUE THREAD

**Issue by aarondfrancis** [handsontable/handsontable#7288]
### Description
Continuing my upgrade from 8.0.0-beta1 to 8.0.0 I've run into an issue. Previously we were able to append rows with every visible column set to `null`, as long as the object had other data present (usually a unique identifier and some meta information). Now with 8.0.0 we're no longer able to do that. 

I understand there are new `setSourceData*` methods but I'm not able to get those to append & show data, even after calling `render`.

I don't necessarily need the old functionality restored, but I do need a way to append a "visually empty" row.

### Steps to reproduce
https://jsfiddle.net/8qn9tyuj/2/
1. Run the fiddle, note the blank row 2.
2. Comment out the beta1 script, uncomment the 8.0.0 release script
3. Run again, notice no row 2 and a console error.

### Your environment
* Handsontable version: 8.0.0
* Browser Name and version: Chrome
* Operating System: OS X

**Comment by wojciechczerniak** [comment]
It's not about the `null`, if you add a value Handsotnable will ignore it as well. Next to the source `data` we're keeping a log of rows and columns, it tracks the length of the dataset when it's initialized or modified with CRUD operations. Methods like `setDataAtCell`, `setDataAtRowProp` and `setSourceDataAtCell` can only alter values of already known cells. `populateFromArray` is the only one that has a mode to insert rows/columns instead overwrite the values. We didn't put that much emphasis on them in the migration guide because they now work exactly as described in the docs but I understand that from your perspective this is an unexpected change that broke your code. CC @scarletfog Maybe we should add this anyway.

Back to the solution. To be sure that Handsontable knows about the new row (creates internal index, meta data etc) it should be created first with `alter`, then filled with data.
```js
hot1.alter('insert_row')
hot1.setDataAtRowProp([
  [1, 'database_id', 2],
  [1, 'name', null],
])
```
https://jsfiddle.net/qpok15wm/

**Comment by aarondfrancis** [comment]
@wojciechczerniak thanks for the help. On your fiddle, I'm still getting an error `Uncaught Error: Assertion failed: Expecting an unsigned number.` which I am also running into locally. Are you getting that?

> but I understand that from your perspective this is an unexpected change that broke your code
Well I was using the beta, so that's pretty expected! No problem.

**Comment by aarondfrancis** [comment]
For reference, here is the "appendData" plugin I landed on from our prior discussions: https://github.com/handsontable/handsontable/issues/6032#issuecomment-575846957

**Comment by wojciechczerniak** [comment]
> For reference, here is the "appendData" plugin I landed on from our prior discussions: [#6032 (comment)](https://github.com/handsontable/handsontable/issues/6032#issuecomment-575846957)

And back then my recommendation was the same, to use alter + setData combination:
https://github.com/handsontable/handsontable/issues/6032#issuecomment-575810896

> @wojciechczerniak thanks for the help. On your fiddle, I'm still getting an error `Uncaught Error: Assertion failed: Expecting an unsigned number.` which I am also running into locally. Are you getting that?

Sorry, my fault. I forgot about #7230. `setDataAtCell` and `setDataAtRowProp` are working on indexes that are visual, here: only some props are defined by `columns`. Since one of the properties of the object isn't, we get an error because we try to set/validate metadata for a non-existent column. I should have used the `setSourceDataAtCell` method. https://jsfiddle.net/ody391hn/

I have not taken under the consideration mixed-use case 🤔

**Comment by aarondfrancis** [comment]
Yup that did it! I have refactored my appendData plugin to work with 8.0.0. It was pretty simple in the end. Thanks for the guidance @wojciechczerniak.


For reference:

```js
import PluginFactory from './plugin-factory';

let AppendData = PluginFactory('appendData');

AppendData.prototype.enable = function () {
    // Add the `appendData` method on to the HoT instance.
    this.hot.appendData = this.appendData.bind(this);
}

AppendData.prototype.appendData = function (data, source) {
    let last = this.hot.countRows();

    let changes = [];

    for (let i = 0; i < data.length; i++) {
        for (let key in data[i]) {
            changes.push([last + i, key, data[i][key]]);
        }
    }

    this.hot.alter('insert_row', data.length);
    this.hot.setSourceDataAtCell(changes, source);
}

export default AppendData;

```
